File: src/main/java/com/sidgate/deck/Card.java
package com.sidgate.deck;

import static java.util.Objects.hash;

public class Card {
	private Suit suit;
	private Rank rank;

	public Card(String name) {
		suit = Suit.getByShortName(name.substring(0, 1));
		rank = Rank.getByShortName(name.substring(1, 2));
	}

	public Suit suit() {
		return suit;
	}

	public Rank rank() {
		return rank;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) {
			return true;
		}
		if (o == null || getClass() != o.getClass()) {
			return false;
		}
		Card card = (Card) o;
		return suit == card.suit && rank == card.rank;
	}

	@Override
	public int hashCode() {
		return hash(suit, rank);
	}

	public boolean isAce() {
		return rank.equals(Rank.ACE);
	}

	@Override
	public String toString() {
		return suit.getShortName() + rank.getShortName();
	}
}

File: src/main/kotlin/sidgate/deck/Card.kt
package sidgate.deck


data class Card(val suit: Suit, val rank: Rank) {
    companion object {
        fun of(name: String) = Card(Suit.getByShortName(name.substring(0, 1)), Rank.getByShortName(name.substring(1, 2)))
    }

    val isAce = rank == Rank.ACE

    override fun toString() = suit.shortName + rank.shortName
}